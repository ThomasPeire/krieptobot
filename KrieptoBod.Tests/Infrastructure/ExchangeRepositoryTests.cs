using KrieptoBod.Infrastructure.Exchange;
using KrieptoBod.Model;
using Moq;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace KrieptoBod.Tests.Infrastructure
{
    public class Tests
    {
        private readonly Mock<IExchangeService> _exchangeService = new Mock<IExchangeService>();

        [SetUp]
        public void Setup()
        {
            //Todo use mock data
            _exchangeService
                .Setup(service => service.GetBalanceAsync())
                .Returns(Task.FromResult(new List<Balance>().AsEnumerable()));
            _exchangeService
                .Setup(service => service.GetCandlesAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<int>(), It.IsAny<DateTime>(), It.IsAny<DateTime>()))
                .Returns(Task.FromResult(new List<Candle>().AsEnumerable()));

            _exchangeService
                .Setup(service => service.GetMarketsAsync())
                .Returns(Task.FromResult(new List<Market>().AsEnumerable()));

            _exchangeService
                .Setup(service => service.GetMarketAsync(It.IsAny<string>()))
                .Returns(Task.FromResult(new Market()));

            _exchangeService
                .Setup(service => service.GetAssetsAsync())
                .Returns(Task.FromResult(new List<Asset>().AsEnumerable()));

            _exchangeService
                .Setup(service => service.GetAssetAsync(It.IsAny<string>()))
                .Returns(Task.FromResult(new Asset()));

            _exchangeService
                .Setup(service => service.GetTradesAsync(It.IsAny<string>(), It.IsAny<int>(), It.IsAny<DateTime>(), It.IsAny<DateTime>(), It.IsAny<Guid>(), It.IsAny<Guid>()))
                .Returns(Task.FromResult(new List<Trade>().AsEnumerable()));

            _exchangeService
                .Setup(service => service.GetOrderAsync(It.IsAny<string>(), It.IsAny<Guid>()))
                .Returns(Task.FromResult(new Order()));

            _exchangeService
                .Setup(service => service.GetOrdersAsync(It.IsAny<string>(), It.IsAny<int>(), It.IsAny<DateTime>(), It.IsAny<DateTime>(), It.IsAny<Guid>(), It.IsAny<Guid>()))
                .Returns(Task.FromResult(new List<Order>().AsEnumerable()));

            _exchangeService
                .Setup(service => service.GetOpenOrderAsync(It.IsAny<string>()))
                .Returns(Task.FromResult(new Order()));

        }

        [Test]
        public async Task GetBalance_ShouldReturn_Balance()
        {
            var exchangeRepository = new ExchangeRepository(_exchangeService.Object);

            var result = await exchangeRepository.GetBalanceAsync();

            Assert.That(result, Is.Not.Null);
        }

        [Test]
        public async Task GetCandles_ShouldReturn_Candles()
        {
            var exchangeRepository = new ExchangeRepository(_exchangeService.Object);

            var result = await exchangeRepository.GetCandlesAsync("", "", 0, DateTime.Now, DateTime.Now);

            Assert.That(result, Is.Not.Null);
        }

        [Test]
        public void GetMarkets_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetMarketsAsync(), Throws.Exception);
        }

        [Test]
        public void GetMarket_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetMarketAsync(""), Throws.Exception);
        }

        [Test]
        public void GetAssets_ShouldThrow_NotImplementedException()
        {
            Assert.That(async() => await new ExchangeRepository(_exchangeService.Object).GetAssetsAsync(), Throws.Exception);
        }

        [Test]
        public void GetAsset_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetAssetAsync(""), Throws.Exception);
        }

        [Test]
        public void GetTrades_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetTradesAsync(""), Throws.Exception);
        }

        [Test]
        public void GetOrders_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetOrdersAsync(""), Throws.Exception);
        }

        [Test]
        public void GetOrder_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetOrderAsync("", new Guid()), Throws.Exception);
        }

        [Test]
        public void GetOpenOrder_ShouldThrow_NotImplementedException()
        {
            Assert.That(async () => await new ExchangeRepository(_exchangeService.Object).GetOpenOrderAsync(""), Throws.Exception);
        }
    }
}