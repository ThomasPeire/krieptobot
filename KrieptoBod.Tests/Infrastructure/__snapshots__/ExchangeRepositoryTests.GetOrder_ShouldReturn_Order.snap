{
  "OrderId": "4a7bd126-2d21-4918-96dc-0c8f51760a0b",
  "Market": "BTC-EUR",
  "Created": "0001-01-02T20:52:24.9931466",
  "Updated": "0001-01-02T20:52:24.9931466",
  "Status": "filled",
  "Side": "sell",
  "OrderType": "limit",
  "Amount": 241469.0,
  "AmountRemaining": 0.0,
  "Price": 45928.0,
  "AmountQuote": 0.0,
  "AmountQuoteRemaining": 0.0,
  "OnHold": 0.0,
  "OnHoldCurrency": "BTC",
  "TriggerPrice": 0.0,
  "TriggerAmount": 0.0,
  "TriggerType": null,
  "TriggerReference": null,
  "FilledAmount": 241469.0,
  "FilledAmountQuote": 11090188232.0,
  "FeePaid": 17188232.0,
  "FeeCurrency": "EUR",
  "Fills": [
    {
      "Id": "c9d0abf0-9377-4b1a-ab06-a840f730dc43",
      "Timestamp": "0001-01-02T20:52:25.1035833",
      "Amount": 241469.0,
      "Price": 45928.0,
      "Taker": false,
      "Fee": 17188232.0,
      "FeeCurrency": "EUR",
      "Settled": true
    }
  ],
  "SelfTradePrevention": "decrementAndCancel",
  "Visible": false,
  "TimeInForce": "GTC",
  "PostOnly": false,
  "DisableMarketProtection": false
}
